### Q1. What is the difference between Ordinal Encoding and Label Encoding? Provide an example of when you might choose one over the other.

Ordinal Encoding and Label Encoding are two commonly used techniques in feature encoding for categorical variables.

Ordinal Encoding assigns an integer value to each category in a categorical variable based on their order or rank. For example, if we have a categorical variable for education level with categories 'High School', 'Bachelor's Degree', and 'Master's Degree', we can assign 'High School' the value 1, 'Bachelor's Degree' the value 2, and 'Master's Degree' the value 3.

Label Encoding, on the other hand, assigns a unique integer value to each category in a categorical variable. For example, if we have a categorical variable for color with categories 'Red', 'Green', and 'Blue', we can assign 'Red' the value 1, 'Green' the value 2, and 'Blue' the value 3.

In some cases, the order of categories in a categorical variable may not be significant, and so Label Encoding would be a more appropriate choice. For example, in a dataset of customer demographic data, the 'gender' feature would be best encoded using Label Encoding because there is no inherent order or ranking of the categories 'Male' and 'Female'.

In other cases, the order of categories in a categorical variable may be meaningful, and so Ordinal Encoding would be a more appropriate choice. For example, in a dataset of student grade levels, the 'grade' feature would be best encoded using Ordinal Encoding because there is a clear order to the categories 'Freshman', 'Sophomore', 'Junior', and 'Senior'.

### Q2. Explain how Target Guided Ordinal Encoding works and provide an example of when you might use it in a machine learning project.

Target Guided Ordinal Encoding is a technique used to encode categorical variables based on their relationship with the target variable in a supervised machine learning problem. The basic idea is to assign a numerical value to each category of the categorical variable based on the mean or median target value for that category. The categories with the highest target value are assigned the highest numerical value, and the categories with the lowest target value are assigned the lowest numerical value.

For example, let's say we have a dataset of customer information for a bank, including a categorical variable "education" with categories "high school", "college", and "graduate school", and a target variable indicating whether or not the customer defaulted on a loan. To perform Target Guided Ordinal Encoding, we would group the data by each category of "education" and calculate the mean or median target value for each group. We would then assign a numerical value to each category based on its mean or median target value. The category with the highest target value would be assigned the highest numerical value, and the category with the lowest target value would be assigned the lowest numerical value.

In a machine learning project, Target Guided Ordinal Encoding can be used when the categorical variable has a strong relationship with the target variable and the goal is to improve the predictive power of the model. For example, if we are building a model to predict customer loan default, the "education" variable may be a good candidate for Target Guided Ordinal Encoding, as it is likely to be a strong predictor of loan default. By encoding the variable based on its relationship with the target, we may be able to improve the accuracy of our model. However, it is important to note that Target Guided Ordinal Encoding can lead to overfitting if not used carefully, and should be used in conjunction with other encoding techniques and feature selection methods.

### Q3. Define covariance and explain why it is important in statistical analysis. How is covariance calculated?

Covariance is a statistical measure that quantifies the extent to which two random variables vary together. In other words, covariance measures how much two variables move in the same direction. A positive covariance indicates that the two variables tend to move together, while a negative covariance suggests that the two variables move in opposite directions.

Covariance is important in statistical analysis because it helps to understand the relationship between two variables. Specifically, it is a useful tool for analyzing the strength and direction of the linear relationship between two variables. Covariance can be used to identify patterns and trends in data, and can be used to help predict future values of a variable based on the values of another variable.

The formula for covariance is:

cov(X,Y) = E[(X - E[X])(Y - E[Y])]

where X and Y are the two random variables, E[X] and E[Y] are the expected values of X and Y, and E[(X - E[X])(Y - E[Y])] is the expected value of the product of the deviations of X and Y from their respective means.

In practice, covariance can be calculated by computing the product of the deviations of each observation from their respective means and then dividing by the total number of observations minus one. The resulting value provides a measure of the degree to which two variables are related. However, it should be noted that covariance is affected by the scale of the variables, which can make it difficult to compare the strength of the relationship between variables with different units of measurement.

In [1]:
#Example
import seaborn as sns
df=sns.load_dataset('tips')
df.cov()

/tmp/ipykernel_119/2792921673.py:3: FutureWarning: The default value of numeric_only in DataFrame.cov is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.cov()


total_bill       tip      size
total_bill   79.252939  8.323502  5.065983
tip           8.323502  1.914455  0.643906
size          5.065983  0.643906  0.904591

### Q4. For a dataset with the following categorical variables: Color (red, green, blue), Size (small, medium, large), and Material (wood, metal, plastic), perform label encoding using Python's scikit-learn library. Show your code and explain the output.

In [2]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder

# Define the data as a list of lists
data = [['red', 'small', 'wood'],
        ['green', 'medium', 'metal'],
        ['blue', 'large', 'plastic'],
        ['red', 'small', 'plastic']]

# Define the column names
columns = ['Color', 'Size', 'Material']

# Create a DataFrame
df = pd.DataFrame(data, columns=columns)

# Print Dataframe before encoding
print(f'Dataframe Before Encoding :\n {df}')
print('\n=================================\n')

# Create a LabelEncoder object
le = LabelEncoder()

# Apply label encoding to each column in the DataFrame
for col in df.columns:
    df[col] = le.fit_transform(df[col])

# Print the encoded DataFrame
print(f'Dataframe After Encoding :\n {df}')

Dataframe Before Encoding :
    Color    Size Material
0    red   small     wood
1  green  medium    metal
2   blue   large  plastic
3    red   small  plastic


Dataframe After Encoding :
    Color  Size  Material
0      2     2         2
1      1     1         0
2      0     0         1
3      2     2         1


In the encoded dataset, each categorical variable has been replaced with numerical values. For example, 'red' is encoded as 2, 'green' as 1, and 'blue' as 0 for the 'Color' variable. Similarly, 'small' is encoded as 2, 'medium' as 0, and 'large' as 1 for the 'Size' variable, and 'wood' is encoded as 2, 'metal' as 1, and 'plastic' as 0 for the 'Material' variable.
This encoding is done based on alphabetical order eg. blue = 0 , green = 1 , red = 2
Note that the encoded values have no inherent meaning or order. They are simply numerical representations of the original categorical variables.

### Q5. Calculate the covariance matrix for the following variables in a dataset: Age, Income, and Education level. Interpret the results.

In [4]:
import numpy as np
import pandas as pd

# Setting random seed 
np.random.seed(765)

# Generating synthetic data
n = 1000
age = np.random.randint(low=25,high=60,size=n)
education_level = np.random.choice(['High School','Bachelor','Masters','PhD'],size=n)
income = 1100*age + np.random.normal(loc=0, scale=5000,size=n)

# Storing in dataframe
df = pd.DataFrame(
    {'age':age,
     'education_level':education_level,
     'income':income}
)

df.head()

age education_level        income
0   54         Masters  59028.015536
1   51         Masters  49213.962387
2   29     High School  32020.177216
3   52        Bachelor  63067.339595
4   42     High School  43945.405198

In [6]:
#Apply Ordinal Encoding on Education variable
from sklearn.preprocessing import OrdinalEncoder
encoder = OrdinalEncoder(categories=[['High School','Bachelor','Masters','PhD']])
edu_endoded = encoder.fit_transform(df[['education_level']])
df['education_level']=np.ravel(edu_endoded)
df.head()

age  education_level        income
0   54              2.0  59028.015536
1   51              2.0  49213.962387
2   29              0.0  32020.177216
3   52              1.0  63067.339595
4   42              0.0  43945.405198

In [8]:
#covariance matrix
df.cov()

age  education_level        income
age                 101.174679         0.298671  1.089271e+05
education_level       0.298671         1.226698  3.421286e+02
income           108927.128329       342.128564  1.426932e+08

### Q6. You are working on a machine learning project with a dataset containing several categorical variables, including "Gender" (Male/Female), "Education Level" (High School/Bachelor's/Master's/PhD), and "Employment Status" (Unemployed/Part-Time/Full-Time). Which encoding method would you use for each variable, and why?


In machine learning projects, categorical variables need to be encoded as numerical values in order to be used as input to models. The choice of encoding method depends on the type and nature of the variable. Here's how I would encode the three categorical variables in this scenario:

1. Gender: Since this variable has only two possible values, Male and Female, we can use binary encoding. In this case, we would create a new column called "Gender_Male" and assign a value of 1 to Male and 0 to Female, or vice versa.

2. Education Level: Since this variable has more than two possible values, we can use one-hot encoding. This means creating a new column for each possible value and assigning a value of 1 or 0 to indicate whether that value is present or not. In this case, we would create four new columns called "Education_Level_HighSchool", "Education_Level_Bachelors", "Education_Level_Masters", and "Education_Level_PhD", and assign a value of 1 to the appropriate column for each observation.

3. Employment Status: Since this variable also has more than two possible values, we can use ordinal encoding. This involves assigning a numerical value to each possible value based on its order or level of importance. In this case, we could assign a value of 1 to Unemployed, 2 to Part-Time, and 3 to Full-Time, since Full-Time is likely more important or indicative of a stable income and financial situation compared to the other two options.

The choice of encoding method is important, as it can impact the performance of machine learning models. Using the appropriate encoding method ensures that the information contained in categorical variables is accurately represented and utilized in the model.

### Q7. You are analyzing a dataset with two continuous variables, "Temperature" and "Humidity", and two categorical variables, "Weather Condition" (Sunny/Cloudy/Rainy) and "Wind Direction" (North/South/ East/West). Calculate the covariance between each pair of variables and interpret the results.

In [9]:
import numpy as np
import pandas as pd

# Set seed for reproducibility
np.random.seed(321)

# Generate data
n = 1000
temp = np.random.normal(25, 5, n)
humidity = np.random.normal(60, 10, n)
weather_condition = np.random.choice(['Sunny', 'Cloudy', 'Rainy'], size=n)
wind_direction = np.random.choice(['North', 'South', 'East', 'West'], size=n)

# Create dataframe
df = pd.DataFrame({
    'Temperature': temp, 
    'Humidity': humidity, 
    'Weather Condition': weather_condition, 
    'Wind Direction': wind_direction
})

# Show first few rows
df.head()

Temperature   Humidity Weather Condition Wind Direction
0    25.862597  50.526311             Sunny          South
1    33.177413  55.809608             Sunny          South
2    25.186682  70.091030             Sunny           West
3    20.579252  68.981094             Sunny          South
4    19.284039  78.624127             Rainy           East

In [10]:
#Calculating Covariance Matrix for Numerical Variables only
df.cov()

/tmp/ipykernel_119/2492690765.py:2: FutureWarning: The default value of numeric_only in DataFrame.cov is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.cov()


Temperature    Humidity
Temperature    25.165416    1.610779
Humidity        1.610779  105.612893

The covariance between "Temperature" and "Humidity" is 1.611 , indicating a positive relationship between the two variables. This means that as temperature increases, humidity tends to increase as well. The variances of each variable are shown on the diagonal, with Humidity having a larger variance than Temperature.

To calculate the covariance between the continuous variables and the categorical variables, we can group the data by the categorical variables and calculate the covariance for each group. Here's an example code:

It is important to note that we cannot calculate the covariance between continuous and categorical variables since covariance requires numerical data. Therefore, we cannot interpret the covariance between "Temperature" and "Weather Condition" or between "Humidity" and "Wind Direction". In general, we need to be careful when interpreting covariance and consider the nature of the variables being analyzed.